### PR TITLE
make mono use global mutex instead of named mutex, like in CoreCLR

### LINKED ISSUE #2860: Bug in ExecuteSynchronizedCore on Linux/MacOS + Mono
Hi, I have the following exception:

```
at <unknown> <0xffffffff>
at (wrapper managed-to-native) System.Threading.WaitHandle.WaitOne_internal (System.Threading.WaitHandle,intptr,int,bool) <0xffffffff>
at System.Threading.WaitHandle.WaitOne (System.TimeSpan,bool) <0x0009b>
at System.Threading.WaitHandle.WaitOne (System.TimeSpan) <0x0001d>
at NuGet.Configuration.Settings.ExecuteSynchronizedCore (System.Action) <0x00143>
at NuGet.Configuration.Settings.ExecuteSynchronized (System.Action) <0x00019>
at NuGet.Configuration.Settings..ctor (string,string,bool) <0x0032b>
at NuGet.Configuration.Settings.ReadSettings (string,string,bool) <0x0008a>
at NuGet.Configuration.Settings.LoadUserSpecificSettings (System.Collections.Generic.List`1<NuGet.Configuration.Settings>,string,string,NuGet.Configuration.IMachineWideSettings,bool) <0x00418>
at NuGet.Configuration.Settings.LoadDefaultSettings (string,string,NuGet.Configuration.IMachineWideSettings,bool,bool) <0x0031e>
at NuGet.Configuration.Settings.LoadDefaultSettings (string,string,NuGet.Configuration.IMachineWideSettings) <0x00026>

Native stacktrace:

0   mono-sgen                           0x000000010e7c976a mono_handle_native_sigsegv + 282
1   libsystem_platform.dylib            0x00007fff8e929f1a _sigtramp + 26
2   mono-sgen                           0x000000010eab167f tmp_dir + 5471
3   libsystem_c.dylib                   0x00007fff86e909b3 abort + 129
4   mono-sgen                           0x000000010e96f7f3 monoeg_log_default_handler + 211
5   mono-sgen                           0x000000010e96f702 monoeg_g_logv + 114
6   mono-sgen                           0x000000010e96fb04 monoeg_assertion_message + 356
7   mono-sgen                           0x000000010e951740 own_if_owned + 0
8   mono-sgen                           0x000000010e8b8430 ves_icall_System_Threading_WaitHandle_WaitOne_internal + 96
9   ???                                 0x0000000113feccb4 0x0 + 4630432948
10  mscorlib.dll.dylib                  0x0000000110ac347e System_Threading_WaitHandle_WaitOne_System_TimeSpan + 30
11  ???                                 0x00000001264dffda 0x0 + 4937613274
```

The origin of the exception:

``` cs
// Global: ensure mutex is honored across TS sessions 
using (var mutex = new Mutex(false, $"Global\\{EncryptionUtility.GenerateUniqueToken(fileName)}"))
{
    var owner = false;
    try
    {
        // operations on NuGet.config should be very short lived
        owner = mutex.WaitOne(TimeSpan.FromMinutes(1));
        // decision here is to proceed even if we were not able to get mutex ownership
        // and let the potential IO errors bubble up. Reasoning is that failure to get
        // ownership probably means faulty hardware and in this case it's better to report
        // back than hang
        ioOperation();
    }
```

(see https://github.com/NuGet/NuGet.Client/blob/58bd2dffe9cee8bf62b601f1610df4e9bbb91106/src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs#L1134)

An observation: on CoreCLR+(Linux/MacOS), NuGet uses `_globalMutex` instead of named `mutex` because cross-plat CoreCLR doesn't support named locks. However, on Mono+(Linux/MacOS), NuGet tries to use a named mutex.

Linux version: Ubuntu 14.04
MacOS version: 10.10.5
Mono version: 4.2.3.4
NuGet.Client: 3.4.3

### PR DESCRIPTION
**Pull request by rohit21agrawal**
fixes https://github.com/NuGet/Home/issues/2860

CC: @joelverhagen @alpaix @rrelyea @yishaigalatzer

### CODE REVIEW

File: src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs [hunk 1]
```diff
@@ -1069,59 +1069,24 @@ private void Save()
...
-                }
-                finally
-                {
-                    if (owner)
-                    {
-                        _globalMutex.ReleaseMutex();
-                    }
-                }
+                ExecuteSynchronizedCoreCrossPlat(ioOperation);
+                return;
```
Comment (drewgillies) [inline src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs]: Given that both cases in this if-else end in return, the returns can be removed altogether.

**Comment by drewgillies** [comment]
Apart from above, :shipit: :)

File: src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs (hunk 1 above)
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs]: fixed

File: src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs [hunk 2]
```diff
@@ -1069,71 +1069,87 @@ private void Save()
...
-                {
-                    if (owner)
-                    {
-                        _globalMutex.ReleaseMutex();
-                    }
-                }
+                ExecuteSynchronizedCoreCrossPlat(ioOperation);
             }
+
+            return;
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs]: not needed

**Comment by yishaigalatzer** [comment]
Why do we want a global mutex rather than a named one all the time?

**Comment by rohit21agrawal** [comment]
@yishaigalatzer we can't ifdef the global mutex declaration because we can't detect if we are running against mono at buildtime.

**Comment by alpaix** [comment]
@rohit21agrawal use `Lazy` global mutex?

**Comment by rohit21agrawal** [comment]
@alpaix good idea. implemented.

**Comment by alpaix** [comment]
:ok: :shipit:

**Comment by rohit21agrawal** [comment]
@yishaigalatzer @alpaix 
Discussed the possibility of using ConcurrentUtilities.ExecuteWithFileLockedAsync in place of mutex here. However, since the method is asynchronous , it will take a huge amount of effort to plumb the async calls through  all the methods. Therefore, decided to go with the mutex approach only.

**Comment by rohit21agrawal** [comment]
The correct fix is being tracked through : https://github.com/NuGet/NuGet.Client/pull/725

**Pull request merged**